File: 20. Valid Parentheses/main.cpp
#include "iostream"
#include <stack>
using namespace std;

bool isValid(string s) {
        int length = s.length();
        stack <char> stac ;
        for (int i = 0; i < length; ++i)
        {
        	if (s[i] == ')' || s[i] == ']' || s[i] == '}')
        	{
        		if (stac.empty() == true)
        		{
        			return false;
        		}
        		else if (s[i] == ')' && stac.top() == '(')
        		{
        			stac.pop();
        		}
        		else if (s[i] == ']' && stac.top() == '[')
        		{
        			stac.pop();
        		}
        		else if (s[i] == '}' && stac.top() == '{')
        		{
        			stac.pop();
        		}
        		else
        		{
        			return false;
        		}
        	}
        	else
        	{
        		stac.push(s[i]);
        	}
        }
        return stac.empty();
    }


int main ()
{
	cout<<isValid("()");
	return 0;
}

File: 20. Valid Parentheses/main.py
#! usr/bin/env python
class Solution(object):
	"""docstring for Solution"""
	def isValid(self, s):
		left = ['(', '[', '{']
		right = [')', ']', '}']
		stac = []
		for char_s in s:
			if char_s in left:
				stac.append(char_s)
			else:
				index = right.index(char_s)
				if stac == []:
					return False
				if stac[-1] == left[index]:
					stac.pop()
				else:
					return False
		return True if stac==[] else False


if __name__=='__main__':
	so = Solution()
	print so.isValid(']')
